lenskit-svd/src/main/java/org/lenskit/mf/BPR/BPRMFModelProvider.java

```diff
@@ -0,0 +1,175 @@
+    private final double learningRate;
+    private final double regularization;
+    private final DataAccessObject dao;
+    private final StoppingCondition stoppingCondition;
+    private final TrainingPairGenerator pairGenerator;
+    private final Random rand;
+
+    private HashKeyIndex itemIndex;
+    private HashKeyIndex userIndex;
+
+    @Inject
+    public BPRMFModelProvider(@Transient @Nonnull DataAccessObject dao,
+                              @Transient @Nonnull TrainingPairGenerator pairGenerator,
+                              @FeatureCount int featureCount,
+                              @LearningRate double learningRate,
+                              @RegularizationTerm double regularizaiton,
+                              @Transient StoppingCondition stoppingCondition,
+                              @Transient Random rand) {
+        this.dao = dao;
+        this.pairGenerator = pairGenerator;
+        this.featureCount = featureCount;
+        this.learningRate = learningRate;
+        this.regularization = regularizaiton;
+        this.stoppingCondition = stoppingCondition;
+        this.rand = rand;
+
+        itemIndex = new HashKeyIndex();
+        userIndex = new HashKeyIndex();
+    }
+
+    @Override
+    public MFModel get() {
+        //REVIEW arbitrary size.
+        // This will accumulate BPR-Opt (minus the regularization) and will be negated to make an error.
+        // -30 is arbitrary, but would indicate a _really_ consistently bad prediction (~ p=1*10^-13),
+        // and is therefore a reasonable "max_error".
+        RollingWindowMeanAccumulator optAccum = new RollingWindowMeanAccumulator(1000, -30);
+
+        int userCount = dao.getEntityIds(CommonTypes.USER).size();
+        RealMatrix userFeatures = MatrixUtils.createRealMatrix(userCount, featureCount);
+        initFeatures(userFeatures);
+
+        int itemCount = dao.getEntityIds(CommonTypes.ITEM).size();
+        RealMatrix itemFeatures = MatrixUtils.createRealMatrix(itemCount, featureCount);
+        initFeatures(userFeatures);
+
+        logger.debug("Learning rate is {}", learningRate);
+        logger.debug("Regularization term is {}", regularization);
+
+        logger.info("Building MPR-MF with {} features for {} users and {} items",
+                featureCount, userCount, itemCount);
+
+        TrainingLoopController controller = stoppingCondition.newLoop();
+
+        //REVIEW: because of the nature of training samples (and the point that the BPR paper makes that training
+        // by-item or by-user are not optimal) one "iteration" here will be one training update. This leads to _really_
+        // big iteration counts, which can actually overflow ints!. one suggestion would be to allow the iteration count
+        // controller to accept longs, but I don't know if this will introduce backwards compatibility issues (I imagine
+        // this depends on the robustness of our type conversion in the configuration.
```

Comment:
I was thinking about this, if we want to use any two count system I think the lower level of counts should go into the training pair generator. That gives the flexibility to do exhaustive all pairs generators in a clean way, I'm just not sure I like the idea of a two count system.